src/libs/actions/Policy.js

```diff
@@ -30,7 +75,69 @@ function getPolicySummaries() {
         });
 }
 
+/**
+ * Fetches the policyList from the API and saves a simplified version in Onyx
+ */
+function getPolicyList() {
+    GetPolicyList()
+        .then((data) => {
+            if (data.jsonCode === 200) {
+                const policyDataToStore = _.reduce(data.policyList, (memo, policy) => ({
+                    ...memo,
+                    [`${ONYXKEYS.COLLECTION.POLICY}${policy.id}`]: getSimplifiedMemberList(policy),
+                }), {});
+                Onyx.mergeCollection(ONYXKEYS.COLLECTION.POLICY, policyDataToStore);
+            }
+        });
+}
+
+/**
+ * Merges the passed in login into the specified policy
+ *
+ * @param {String} login
+ * @param {String} welcomeNote
+ * @param {String} policyID
+ */
+function invite(login, welcomeNote, policyID) {
+    const key = `${ONYXKEYS.COLLECTION.POLICY}${policyID}`;
+
+    // Make a shallow copy to preserve original data, and concat the login
+    const policy = _.clone(allPolicies[key]);
+    policy.employeeList = [...policy.employeeList, login];
+
+    // Optimistically add the user to the policy
+    Onyx.set(key, policy);
+
+    // Make the API call to merge the login into the policy
+    Policy_Employees_Merge({
+        employees: JSON.stringify([{email: login}]),
+        welcomeNote,
+        policyID,
+    })
+        .then((data) => {
+            // Save the personalDetails for the invited user in Onyx
+            if (data.jsonCode === 200 && data.personalDetails) {
+                Onyx.merge(ONYXKEYS.PERSONAL_DETAILS, formatPersonalDetails(data.personalDetails));
+                return;
+            }
+
+            // If the operation failed, undo the optimistic addition
+            const policyDataWithoutLogin = _.clone(allPolicies[key]);
+            policyDataWithoutLogin.employeeList = _.without(allPolicies[key].employeeList, login);
+            Onyx.set(key, policyDataWithoutLogin);
```

Comment:
NAB, if we use `Onyx.merge()` this can be shortened to something like...

```js
Onyx.merge(key, {employeeList: _.without(allPolicies[key].employeeList, login)});
```

Reply:
I created a thread in slack about this, I was seeing an issue where `merge` didn't overwrite the newly added value (https://expensify.slack.com/archives/C01GTK53T8Q/p1623391961201700) which is why I used set here.

Reply:
What's the issue exactly? I can't think of a reason why this shouldn't work as we should be replacing array values when they are on objects and not concatenating them. Curious why this is not working..

Reply:
The issue is that calling Onyx.merge on a nested employeeList won't overwrite the value of the employeeList to remove the new login. So if `test@test.com` is set in the employeeList and we call `Onyx.merge(key, {employeeList: _.without(allPolicies[key].employeeList, login)});`, the `test@test.com` login remains in employeeList.

Reply:
> as we should be replacing array values when they are on objects

I don't know about Onyx.merge but Onyx.mergeCollection would not work like that - it will not replace the array but merge it. 
So if you have object with a nested array with 3 elements and pass a delta with nested array of 2 elements - the end result is nested  array with 3 elements

Reply:
_.merge would do the same thing

Reply:
>I don't know about Onyx.merge but Onyx.mergeCollection would not work like that - it will not replace the array but merge it.

Well, we're not discussing `mergeCollection()` in this case, but that behavior is also not consistent with my expectations or testing and again... I wouldn't expect these arrays to get concatenated recursively...

**`_.merge()`**
![Screen Shot 2021-06-15 at 12 24 58 PM](https://user-images.githubusercontent.com/32969087/122131378-ceba2a80-cdd4-11eb-8824-2d6cb11a5660.png)

![Screen Shot 2021-06-15 at 12 33 46 PM](https://user-images.githubusercontent.com/32969087/122132127-fc53a380-cdd5-11eb-8fae-d06348c69bb3.png)

![Screen Shot 2021-06-15 at 12 41 42 PM](https://user-images.githubusercontent.com/32969087/122132770-32455780-cdd7-11eb-8314-6c90a3ce1700.png)

Reply:
Ok I think this was a slight distraction, but pretty interesting. the default behavior of `merge()` seems like not to concat.

https://github.com/lodash/lodash/issues/2872


TL;DR it seems correct to use `Onyx.set()` here (for now) - but we should clarify when `merge()` should or should not be used somewhere. Or update it to use `_.mergeWith()` + customizer like in this example:

https://lodash.com/docs/4.17.15#mergeWith

Reply:
> the default behavior of `merge()` seems like not to concat.

That's what I tried to say as well.

Keep in mind that thought it's possible to use `_.mergeWith` it will only work for local data - you can't istruct `AsyncStorage.merge` and `AsyncStorage.multiMerge` how to merge data - these methods are not configurable and will merge the way `_.merge` works by default - arrays are not concatenated

So if you change how `Onyx.merge` to **replace** arrays, you'd have to make the same change for `Onyx.mergeCollection`, but then you'll be forced to read the entire collection since you can't define ahead of time how data should be merged

Reply:
Yes, I think we'd have to come up with some other solution. Left some thoughts on it in https://github.com/Expensify/react-native-onyx/issues/83 if you have any context or thoughts to add.